crates/bitwarden-vault/src/folder/edit.rs

```diff
@@ -0,0 +1,212 @@
+use bitwarden_api_api::apis::folders_api;
+use bitwarden_core::{key_management::KeyIds, require, ApiError, MissingFieldError};
+use bitwarden_crypto::{CryptoError, KeyStore};
+use bitwarden_error::bitwarden_error;
+use bitwarden_state::repository::{Repository, RepositoryError};
+use thiserror::Error;
+#[cfg(feature = "wasm")]
+use wasm_bindgen::prelude::*;
+
+use crate::{Folder, FolderAddEditRequest, FolderView, VaultParseError};
+
+/// Item does not already exist error.
+#[derive(Debug, thiserror::Error)]
+#[error("Item does not already exist")]
+pub struct ItemDoesNotExistError;
+
+#[allow(missing_docs)]
+#[bitwarden_error(flat)]
+#[derive(Debug, Error)]
+pub enum EditFolderError {
+    #[error(transparent)]
+    ItemDoesNotExist(#[from] ItemDoesNotExistError),
+    #[error(transparent)]
+    Crypto(#[from] CryptoError),
+    #[error(transparent)]
+    Api(#[from] ApiError),
+    #[error(transparent)]
+    VaultParse(#[from] VaultParseError),
+    #[error(transparent)]
+    MissingField(#[from] MissingFieldError),
+    #[error(transparent)]
+    RepositoryError(#[from] RepositoryError),
+    #[error(transparent)]
+    Uuid(#[from] uuid::Error),
+}
+
+pub(super) async fn edit_folder<R: Repository<Folder> + ?Sized>(
+    key_store: &KeyStore<KeyIds>,
+    api_config: &bitwarden_api_api::apis::configuration::Configuration,
+    repository: &R,
+    folder_id: &str,
+    request: FolderAddEditRequest,
+) -> Result<FolderView, EditFolderError> {
+    // Verify the folder we're updating exists
+    repository
+        .get(folder_id.to_owned())
+        .await?
+        .ok_or(ItemDoesNotExistError)?;
+
+    let folder_request = key_store.encrypt(request)?;
+
+    let resp = folders_api::folders_id_put(api_config, folder_id, Some(folder_request))
+        .await
+        .map_err(ApiError::from)?;
+
+    let folder: Folder = resp.try_into()?;
+
+    repository
+        .set(require!(folder.id).to_string(), folder.clone())
```

Comment:
You could slightly simplify this by using the provided folder id:
```suggestion
        .set(folder_id.to_string(), folder.clone())
```

Should we also add a debug assert to make sure the id returned by the api matches the provided one?
```rust
debug_assert_eq!(folder_id, &folder.id.to_string());
```